File: java/src/main/606constructStrFromBinaryTree/solution3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public String tree2str(TreeNode t) {
        if(t == null){
            return "";
        }
        String tmp = helper(t);
        return tmp.substring(1, tmp.length()-1);
    }
    
    private String helper(TreeNode t){
        if(t.left == null && t.right == null){
            return "(" + t.val + ")";
        }
        StringBuilder sb = new StringBuilder("(");
        sb.append(t.val);
        sb.append(t.left == null ? "()" : helper(t.left));
        sb.append(t.right == null ? ""   : helper(t.right));
        sb.append(")");
        return sb.toString();
    }
}

File: python/1379FindNodeInClone/solution_3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        def helper(node):
            if node:
                yield node
                yield from helper(node.left)
                yield from helper(node.right)
        if original is None:
            return None
        return [x[1] for x in zip(helper(original), helper(cloned)) if x[0] == target][0] 
